mesonbuild/wrap/wrap.py

```diff
@@ -462,7 +462,17 @@ def resolve(self, packagename: str, method: str) -> str:
             if not os.path.isdir(self.dirname):
                 raise WrapException('Path already exists but is not a directory')
         else:
-            if self.wrap.type == 'file':
+            # Check first if we have the extracted directory in our cache. This can
+            # happen for example when MESON_PACKAGE_CACHE_DIR=/usr/share/cargo/registry
+            # on distro that ships Rust source code. This can also happen when the git
+            # repository has been cloned into the cache to avoid cloning multiple
+            # times. Copy the tree because we could be applying patches on top.
+            # TODO: In case of a git repo, we could add a worktree instead of
+            # copy_tree().
```

Comment:
As discussed on IRC, this has some pretty bad effects. Cloning would be hardlinked, but then you get the "wrong" remote...

Reply:
We don't currently clone git in our cache. When we'll start doing that we would need to make sure to checkout the correct branch of course.

Reply:
I think Eli just wants the TODO comment updated, and I agree. Maybe just say:

```suggestion
            # TODO: In case of a git repo, we could do something smarter instead of
            # copy_tree().
```

Reply:
Right. It seems odd to have a TODO that recommends a specific resolution which we've already somewhat agreed is "probably the wrong way to do it". Actual investigation into optimizing git repos could result in any of several alternative solutions.

The proposed comment tweak makes sense to me.

Reply:
Reformulated to move the whole git story under TODO and removed the worktree mention.